farmer/src/lib.rs

```diff
@@ -372,11 +369,11 @@ impl Config {
                 parking_lot::Mutex::new(lru::LruCache::new(SEGMENT_COMMITMENTS_CACHE_SIZE)),
             )),
         );
-        let piece_getter = Arc::new(CombinedPieceGetter::new(
+        let farmer_piece_getter = Arc::new(FarmerPieceGetter::new(
```

Comment:
Ideally SDK would not use RPC at all and would disable it outright. `NodeClient` is a trait in farmer specifically such that SDK could communicate with node directly instead of going through RPC server and JSON serialization/deserialization.

Reply:
Yes. The node client passed here does not use rpc and internally calls node’s rpc module directly.

Reply:
@nazar-pc I actually wanted to do what you described here in first place.

But, Some rpc methods like `piece` and `subscribe_archived_segment_header` incorporates custom logic which I would prefer not to maintain from my side. If I do implement it the way it is described the behavior will diverge from the standard rpc behaviour. 

For example,  `subscribe_archived_segment_header` maintains custom data structure and logic to maintain cached segment header. A straightforward implementation here that directly reads those segment header would change the way node interacts with farmer. 

Unless either a) The custom logic is not important and I can directly query node's store without any negative consequences b) I can use the RpcObject directly (i.e without rpc server in front), I believe using it this way is better than writing custom `NodeClient`.

Reply:
> For example, subscribe_archived_segment_header maintains custom data structure and logic to maintain cached segment header. A straightforward implementation here that directly reads those segment header would change the way node interacts with farmer.

The primary reason it exists there is because node and farmer are decoupled and RPC needs to maintain access to archived segments until archived segment is acknowledge. This should largely be non-issue and greatly simplified in case node is available directly. Even if part of the logic needs to be replicated, it will be much simpler and easier to follow.

You need to understand what things are actually doing and evaluate what makes sense. In some cases we might even need to abstract over `NodeClient`. The primary reason we send headers to farmer and then make farmer request pieces is that segment will not fit into RPC message limit. This issue doesn't exist when node and farmer are in the same process, you can send the whole thing efficiently and not require `NodeClient` as such, but that might require additional refactoring on the farmer that you should analyze and propose architecture for.

Reply:
Make sense. Okay. I will open a issue and work towards it.

Reply:
Keep in mind that it is probably not the highest priority right now, just something to keep in mind and get to eventually